eth/downloader/whitelist/service_test.go

```diff
@@ -30,3 +33,64 @@ func TestWhitelistCheckpoint(t *testing.T) {
 	s.DequeueCheckpointWhitelist()
 	assert.Equal(t, len(s.GetCheckpointWhitelist()), 10, "expected 10 items in whitelist")
 }
+
+// TestIsValidChain checks che IsValidChain function in isolation
+// for different cases by providing a mock fetchHeadersByNumber function
+func TestIsValidChain(t *testing.T) {
+	t.Parallel()
+
+	s := NewMockService(10)
+
+	// case1: no checkpoint whitelist, should consider the chain as valid
+	res, err := s.IsValidChain(nil, nil)
+	assert.Equal(t, res, true, "expected chain to be valid")
+	assert.NilError(t, err, "expected no error")
+
+	// add checkpoint entries and mock fetchHeadersByNumber function
+	s.ProcessCheckpoint(uint64(0), common.Hash{})
+	s.ProcessCheckpoint(uint64(1), common.Hash{})
+	assert.Equal(t, len(s.GetCheckpointWhitelist()), 2, "expected 2 items in whitelist")
+
+	// create a false function, returning absolutely nothing
+	falseFetchHeadersByNumber := func(number uint64, amount int, skip int, reverse bool) ([]*types.Header, []common.Hash, error) {
+		return nil, nil, nil
+	}
+
+	// create a mock function, returning a the required header
+	fetchHeadersByNumber := func(number uint64, amount int, skip int, reverse bool) ([]*types.Header, []common.Hash, error) {
+		hash := common.Hash{}
+		header := types.Header{Number: big.NewInt(0)}
+		switch number {
+		case 0:
+			return []*types.Header{&header}, []common.Hash{hash}, nil
+		case 1:
+			header.Number = big.NewInt(1)
+			return []*types.Header{&header}, []common.Hash{hash}, nil
+		case 2:
+			header.Number = big.NewInt(1) // sending wrong header for misamatch
+			return []*types.Header{&header}, []common.Hash{hash}, nil
+		default:
+			return nil, nil, errors.New("invalid number")
+		}
+	}
+
+	// case2: false fetchHeadersByNumber function provided, should consider the chain as valid
```

Comment:
Could you make this cases as separate subtests using t.Run and table tests - https://go.dev/blog/subtests ?
Or this cases are part of one bigger single test?

Reply:
Oh I see, wasn't aware of this thing. Yes, my intention was to have these cases as sub-tests. Will make necessary changes.

Reply:
Great!  Also please check how to use t.Parallel() with subtests.